### PR TITLE
Marionette testing and network simulation

### PR DESCRIPTION
**Pull request by Ackar**
This adds Marionette testing via the `npm run-script marionette` command, and also a script to change the local network setting between the proxy and the client.
You can use it with `npm run-script networksimulation EDGE|HSDPA|LTE|default` (use `default` to restore default config).

### CODE REVIEW

File: config/test/test.yml [hunk 1]
```diff
@@ -29,21 +29,19 @@ metrics:
...
     enabled: false
 
-  phantom:
-    host: '127.0.0.1'
-    port: 3000
-
-  proxy:
-    host: '127.0.0.1'
-    port: 44044
+  firefoxPath: null
```
Comment (eamsen) [inline config/test/test.yml]: Please add a comment about what this is used for.

File: test/helper/networksimulator.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
+'use strict';
+
+var exec = require('child_process').spawn;
+var read = require('read');
+var config = require('config');
+
+// Network profiles
+var networkConfigs = exports.networkConfigs = {
```
Comment (eamsen) [inline test/helper/networksimulator.js]: Is that a constant? Should be in all caps, if so.

File: test/helper/networksimulator.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  EDGE: {
+    latency: 300,
```
Comment (eamsen) [inline test/helper/networksimulator.js]: Please add a comment about the units used in this config.

File: test/helper/networksimulator.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    bandwidth: 200,
+  },
+  HSDPA: {
+    latency: 100,
+    bandwidth: 1000,
```
Comment (eamsen) [inline test/helper/networksimulator.js]: This looks more like HSPA.

File: test/helper/networksimulator.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  },
+  LTE: {
+    latency: 25,
+    bandwidth: null,
+  },
+  default: {
```
Comment (eamsen) [inline test/helper/networksimulator.js]: There are many different HSPA and LTE versions, maybe we should just have 2G, 3G and 4G as more generalized connection type classes instead?

Based on [this book](http://chimera.labs.oreilly.com/books/1230000000545) we would then have a mapping like this:
- 2G: 100 Kbit/s, 1000 ms
- 3G: 512 Kbit/s, 500 ms
- 4G: 1024 Kbit/s, 100ms

These would be the worst case numbers, alternatively we could soften them up a little.

File: test/helper/networksimulator.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    if (callback) {
+      callback();
+    }
+  });
+}
+
+// Linux functions
+
+var LinuxFunctions = {};
+LinuxFunctions.applyProfile = function(prof, callback) {
```
Comment (eamsen) [inline test/helper/networksimulator.js]: We're using prototype-based "class" functions throughout the code, I think we should keep it consistent.

File: test/helper/networksimulator.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  if (!prof.latency && !prof.bandwidth) {
+    if (callback) {
+      callback();
+    }
+    return;
+  }
+
+  sudoExec(['tc', 'qdisc', 'add', 'dev', 'lo', 'root', 'handle', '1:', 'prio'],
+  function() {
```
Comment (eamsen) [inline test/helper/networksimulator.js]: I assume the linter doesn't like the better looking indentation style?

File: test/helper/networksimulator.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+// Run the given command with sudo and ask for password if necessary.
+function sudoExec(cmd, callback) {
+  var child = exec('sudo', ['-S', '-p', '#prompt#'].concat(cmd),
+                   { stdio: 'pipe' });
+
+  child.stderr.on('data', function(data) {
+    if (data.toString().indexOf('#prompt#') > -1) {
+      console.log('Root password needed for network simulation.');
+      read({ prompt: 'Password:', silent: true }, function(error, result) {
```
Comment (eamsen) [inline test/helper/networksimulator.js]: That's convenient for manual usage (which is the current use case), but would not be appropriate for automatic execution. We could just fail and print an error message about insufficient rights instead. @snorp what do you think?

File: test/helper/networksimulator.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  });
+};
+
+BSDFunctions.restoreDefaultConfig = function(callback) {
+  sudoExec(['ipfw', 'delete', '100'], function() {
+    sudoExec(['ipfw', 'pipe', '1', 'delete'], callback);
+  });
+};
+
+var fcts = null;
```
Comment (eamsen) [inline test/helper/networksimulator.js]: A more descriptive name would be better.

File: test/helper/networksimulator.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    args = args.concat(['bw', prof.bandwidth + 'Kbits/s']);
+  }
+
+  sudoExec(args, function() {
+    console.log('Applying filters on packets coming from localhost:%d',
+                config.proxy.port);
+    sudoExec(['ipfw', 'add', '100', 'pipe', '1', 'ip', 'from', 'localhost',
+             config.proxy.port, 'to', 'localhost', 'in'], callback);
+  });
+};
```
Comment (eamsen) [inline test/helper/networksimulator.js]: There is a lot of redundancy in the `applyProfile` functions, the only thing that changes are the `args` strings and CLI commands. Let's just pass those to a generalized function instead.

File: test/helper/testHelper.js [hunk 11]
```diff
@@ -25,10 +66,12 @@ function unloadModule(module) {
...
   SpdyProxy = require('../../lib/proxy');
   proxy = new SpdyProxy(config);
-  proxy.listen(config.test.proxy.port);
+  proxy.listen(config.proxy.port);
+  pacServer.listen(config.pac.port);
 };
 
-exports.localAddress = 'http://127.0.0.1:' + config.test.localServer.port + '/';
+exports.localAddress = 'http://127.0.0.1:' +
+                       config.test.localServer.port + '/';
```
Comment (eamsen) [inline test/helper/testHelper.js]: Why the line break?

File: test/helper/testHelper.js [hunk 12]
```diff
@@ -133,18 +182,19 @@ exports.readResponse = function(response, cb) {
...
+// Convert byte size to size string with appropriate unit.
+exports.getPrettySize = function(size) {
+  var units = ['B', 'KB', 'MB', 'GB'];
+  var i = 0;
 
-  fs.close(fd);
+  while (size > 1024) {
+    size /= 1024;
+    i++;
+  }
```
Comment (eamsen) [inline test/helper/testHelper.js]: That gives `undefined` for `TB` and up. The condition should be `while (size > 1024 && i + 1 < units.length)`.

File: test/perfomance/marionette.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+//  'http://www.google.com/',
+  'http://news.google.com/news',
+  'http://www.cnn.com/',
+//  'http://www.nytimes.com/',
+//  'http://espn.go.com',
+//  'http://imgur.com',
+//  'http://cbs.com',
+//  'http://nbc.com',
+//  'http://abc.com',
+];
```
Comment (eamsen) [inline test/perfomance/marionette.js]: Can we have the top 100 (or more) HTTP sites out of Alexa's top 500 instead?
The performance test can take some time, but should give us some statistically relevant numbers.

File: test/perfomance/marionette.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+function getPercentage(a, b) {
+  return (100 - (b * 100) / a).toFixed(2);
+}
```
Comment (eamsen) [inline test/perfomance/marionette.js]: I've seen that one before, think about moving it out to `utils.js`.

**Comment by eamsen** [comment]
Done with the first pass, lots of nice stuff in there, needs some minor adjustments.
The network simulator could be interesting as a stand-alone module, too.

Also, please update the readme on how to run the tests.

**Comment by Ackar** [comment]
Updated.

File: test/helper/networksimulator.js [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    setupCmds[0] = setupCmds[0].concat(['bw', bandwidth + 'Kbits/s']);
+  }
+
+  var cleanupCmds = [['ipfw', 'delete', '100'],
+    ['ipfw', 'pipe', '1', 'delete']];
+
+  return { setup: setupCmds, cleanup: cleanupCmds };
+}
+
+var getCmds = null;
```
Comment (eamsen) [inline test/helper/networksimulator.js:115]: `getCommands`

File: test/helper/networksimulator.js [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+  child.on('close', function() {
+    if (callback) {
+      callback();
+    }
+  });
+}
+
+// chain the sudoExec calls
+function execSudoCmds(cmds, callback, currentCmd) {
```
Comment (eamsen) [inline test/helper/networksimulator.js:60]: Let's be consistent with abbreviations, I think it's OK to just use the full word here.

**Pull request merged**